src/base.jl

```diff
@@ -1 +1,38 @@
+
 @functor Base.RefValue
+
+functor(::Type{<:Base.ComposedFunction}, x) = (outer = x.outer, inner = x.inner), y -> Base.ComposedFunction(y.outer, y.inner)
+
+using Base.Iterators
+# The reason for these is that calling `Iterators.cycle(data) |> gpu` in Flux should just work.
+
+@functor Iterators.Cycle
+@functor Iterators.Drop (xs,)
+@functor Iterators.Enumerate
+@functor Iterators.Flatten
+@functor Iterators.ProductIterator
+@functor Iterators.PartitionIterator (c,)
+@functor Iterators.Repeated
+@functor Iterators.Reverse
+@functor Iterators.Take (xs,)
+@functor Iterators.Zip
+
+using LinearAlgebra
+# The reason for these is to let W and W' be seen as tied weights in Flux models.
+# But the problem is that the gradient of W' may be a Matrix, so the trees don't match,
+# so we will call `lazywrap` instead of walking:
+
+functor(::Type{<:Adjoint}, x) = (parent = parent(x),), y -> adjoint(only(y))
+lazywrap(x::Adjoint) = parent(x), adjoint
+
+functor(::Type{<:Transpose}, x) = (parent = parent(x),), y -> transpose(only(y))
+lazywrap(x::Transpose) = parent(x), transpose
```

Comment:
This is #28 and should perhaps be pulled out. 

From https://github.com/FluxML/Optimisers.jl/pull/42 I think the right pattern is actually more like this -- calling `parent(x)` is not safe, and we may want to transpose things which aren't matrix-like.
```
Functors.functor(::Type{<:Transpose}, x) = (parent = _transpose(x),), y -> _transpose(only(y))
_transpose(x) = transpose(x)
_transpose(x::NamedTuple{(:parent,)}) = x.parent  # "structural" gradient
# _transpose(bc::Broadcast.Broadcasted{S}) where S = Broadcast.Broadcasted{S}(map(_transpose, bc.args)) do args...
#     transpose(bc.f(map(transpose, args)...))
# end
_transpose(bc::Broadcast.Broadcasted) = transpose(Broadcast.materialize(bc))
```